sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -200,18 +200,18 @@ class Analyzer(
   val postHocResolutionRules: Seq[Rule[LogicalPlan]] = Nil
 
   lazy val batches: Seq[Batch] = Seq(
+    Batch("Substitution", fixedPoint,
```

Comment:
Why we have to execute the batch Substitution before the batch ResolveHints? What is the root cause? I am unable to tell it from either PR description or the code comments. Could any of you explain it? 

cc @maryannxue

Reply:
I don't know which PR causes this regression problem since there is a big gap between 2.4 and 3.0 about CTE. I am not familiar with this part.
I fixed this by debuging. I found only the `child of CTE` could enter into the rule of `ResolveCoalesceHints`
For the above test case:
```
CTE [cte]
:  +- 'SubqueryAlias `cte`
:     +- 'UnresolvedHint REPARTITION, [3]
:        +- 'Project [*]
:           +- 'UnresolvedRelation `t`
+- 'Project [*]
   +- 'UnresolvedRelation `cte`
```
Only this child `Project` branch could enter into the rule of `ResolveCoalesceHints`
```
+- 'Project [*]
   +- 'UnresolvedRelation `cte`
```